### PR TITLE
Add Alchemist Server

### PR DESCRIPTION
**Pull request by SirDoctors**
This was a lot of work and theres still things to be done, however it
should be good for now.

Rocket is the backend for the web server with serde for json, it takes
requests at "http://<domain>/api/v0/". There are three endpoints ATM:
    - ``/`` for the index, takes a GET
    - ``/version["/", "/client", "server"]`` returns the current
    versions of the client and server on GitHub respectivly, or both if not
    specified.
    - ``/packages`` Takes POST request with the package[s], distro name,
    the name of the client and version of the client.

This is a pretty big change to how Alchemist will work and is a on going
progress.

### CODE REVIEW

**Comment by SirDoctors** [comment]
Fixes: #77 
On the block to be fixed are: #75 #26 #79 

Until the issues mentioned above are fixed this shouldn't be considered for merging/

**Comment by SirDoctors** [comment]
This is also waiting on #81 for Travis support for these changes.

File: alchemist_server/src/main.rs [hunk 1]
```diff
@@ -1,3 +1,204 @@
...
+/// ```
+/// Addtionly you can GET '''/version/client''' & ```/version/server```
+/// respectivly to get only one or the other in the response.
+///
+/// ### Package
+/// ``POST /package``
+/// Takes a JSON POST with four variables ``package``, ``distro``, ``client``, ``client_version``.
+/// ``Package`` should be what the user specified for the package to search for.
+/// ``distro`` is the distro that the user is currently using.
+/// ``client``[ ``name`` & ``version`` ] are for current client accessing the API, used for analics & to
```
Comment (mgattozzi) [inline alchemist_server/src/main.rs]: analics?

File: alchemist_server/src/main.rs [hunk 2]
```diff
@@ -1,3 +1,204 @@
...
+/// "client": {
+///     "name": "alchemist",
+///     "version": "0.0.4"
+///     }
+/// }
+/// ```
+/// Returns the package to be installed for the distro supplied in the POST request.
+/// ```json
+/// {
+/// "package": "libgcc"
```
Comment (mgattozzi) [inline alchemist_server/src/main.rs]: I would think it would be:

```json
{
    "package": ["libgcc"]
}
```

File: alchemist_server/src/main.rs [hunk 3]
```diff
@@ -1,3 +1,204 @@
...
+
+use self::alchemist_server::*;
+use self::alchemist_server::models::*;
+use rocket_contrib::{JSON, Value};
+use rocket::State;
+use std::collections::HashSet;
+use std::string::String;
+
+// The serde structs used for Serializing & Deserializing JSON structures.
+#[feature(serde_derive)]
```
Comment (mgattozzi) [inline alchemist_server/src/main.rs]: I don't think you need this anymore

File: alchemist_server/src/main.rs [hunk 4]
```diff
@@ -1,3 +1,204 @@
...
+/// {
+/// "package": "libgcc"
+/// }
+/// ```
+
+extern crate rocket;
+extern crate serde;
+#[macro_use] extern crate rocket_contrib;
+#[macro_use] extern crate serde_derive;
+#[macro_use] extern crate serde_derive;
```
Comment (mgattozzi) [inline alchemist_server/src/main.rs]: This is a duplicate line

File: alchemist_server/src/main.rs [hunk 5]
```diff
@@ -1,3 +1,204 @@
...
+            client: "0.0.4"
+        }, Server_Version {
+            server: "0.0.1"
+        })
+    }
+}
+
+// The package post, allows you to POST and get the correct package for your picture
+#[post("/package", format = "application/json", data = "<message>")]
+fn package(pac: JSON<Package_Incoming>) -> JSON<Package> {
```
Comment (mgattozzi) [inline alchemist_server/src/main.rs]: Maybe an Option would be better?

JSON\<Option\<Package\>\>

File: alchemist_server/src/main.rs [hunk 6]
```diff
@@ -1,3 +1,204 @@
...
+    let mut hashset_pac = HashSet::new();
+    let v = pac.package;
+
+    for p in v {
+        hashset_pac.insert(p);
+    }
+
+    let distro_packages = convert_to_distro(hashset_pac, pac.distro);
```
Comment (mgattozzi) [inline alchemist_server/src/main.rs]: You should be able to turn this into a Vec by doing something like:

convert_to_distro(hashset_pac, pac.distro).into_iter().collect::<Vec<String>>();

That way you won't need to create an alchem or or alchem_fail Vec to push things into.

File: alchemist_server/src/main.rs [hunk 7]
```diff
@@ -1,3 +1,204 @@
...
+            package: alchem_fail
+        })
+    }
+}
+
+// 404 page
+#[error(404)]
+fn not_found(request: &Request) -> content::HTML<String> {
+    let html = match request.content_type() {
+        Some(ref ct) if !ct.is_json() => {
```
Comment (mgattozzi) [inline alchemist_server/src/main.rs]: Wait you can actually match on a pattern AND do a boolean check on the same line as part of the match condition? That's dope. I didn't know that

File: alchemist_server/src/models.rs [hunk 8]
```diff
@@ -0,0 +1,35 @@
+/// Struct for a row in the Alchemist package table
+#[derive(PartialEq,PartialOrd,Eq,Ord,Hash,Debug,Queryable)]
+pub struct Packages {
```
Comment (mgattozzi) [inline alchemist_server/src/models.rs]: If it's just a struct for the row then maybe Package would be better.

File: alchemist_server/.env [hunk 9]
```diff
@@ -0,0 +1 @@
+DATABASE_URL=postgres://benjamin:nope_TM@localhost/alchemist
```
Comment (SirDoctors) [inline alchemist_server/.env]: Thats there more for development purposed, probs going to through in a toml config for that (and possibly other) settings. Will remove it next push though as it wasn't intended to be added.

File: alchemist_server/alchemy/arch.rs [hunk 10]
```diff
@@ -1,261 +0,0 @@
-use std::process::Command;
```
Comment (SirDoctors) [inline alchemist_server/alchemy/arch.rs]: It is, you moved all the files into the server.

File: alchemist_server/migrations/20170317144416_packages/up.sql [hunk 11]
```diff
@@ -0,0 +1,14 @@
+CREATE TABLE packages (
+    id SERIAL PRIMARY KEY,
+    name VARCHAR NOT NULL,
```
Comment (SirDoctors) [inline alchemist_server/migrations/20170317144416_packages/up.sql]: Limits the string size, if need be it can be changed but it shouldn't hurt anything (its a char limit)

File: alchemist_server/migrations/20170317144416_packages/up.sql [hunk 12]
```diff
@@ -0,0 +1,14 @@
...
+    Arch VARCHAR NOT NULL,
```
Comment (SirDoctors) [inline alchemist_server/migrations/20170317144416_packages/up.sql]: They already are locally following changes I made, will be fixed next push.

File: alchemist_server/src/db.rs [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+
+use diesel::prelude::*;
+use diesel::pg::PgConnection;
+use dotenv::dotenv;
+use models::Packages;
+use std::env;
+use std::string::String;
+use std::collections::HashSet;
+
+
```
Comment (SirDoctors) [inline alchemist_server/src/db.rs]: Whoops, now fixed locally.

File: alchemist_server/src/db.rs [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+/// # Examples
+///
+/// ```
+/// let mut packages: HashSet<&str> = HashSet::new()
+/// packages.insert("sudo")
+/// packages.insert("postgresql")
+/// let queryed = pack_query(packages);
+/// ```
+///
+pub fn pack_query(input_packages: HashSet<String>) -> HashSet<Packages> {
```
Comment (SirDoctors) [inline alchemist_server/src/db.rs]: I have a test written for incoming JSON requests, however I don't think it covers this function specifically.

File: alchemist_server/src/db.rs [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+        for j in results {
+            output.insert(j);
+        }
+    }
+
+    output
+}
+
+/// Convert package names from other distros to the one being run by the user currently
+fn convert_to_distro(input_packages: HashSet<String>, distro: String) -> HashSet<String> {
```
Comment (SirDoctors) [inline alchemist_server/src/db.rs]: Yeah totally! That's just a temp case to ensure it works.

File: alchemist_server/src/db.rs [hunk 16]
```diff
@@ -0,0 +1,118 @@
...
+    let results = pack_query(input_packages);
+    let mut pac_converted: HashSet<String> = HashSet::new();
+
+    // All querys will either be a string or '' in the db allowing us to
+    // use is_empty()
+    // Finds out what distro is used and inserts the proper conversions
+    // into the HashSet to be returned from the function
+    match distro {
+        Ubuntu => {
```
Comment (SirDoctors) [inline alchemist_server/src/db.rs]: Its an import from diesel

File: alchemist_server/src/db.rs [hunk 17]
```diff
@@ -0,0 +1,118 @@
...
+        }
+        FreeBSD => {
+            for i in results {
+                if !i.FreeBSD.is_empty() {
+                    pac_converted.insert(i.FreeBSD);
+                }
+            }
+        }
+        _ => {
+            pac_converted.insert(String::new());
```
Comment (SirDoctors) [inline alchemist_server/src/db.rs]: Its a left over that I saw in the original DB code left in the old client. I'll double check to see if its needed.

File: alchemist_server/src/main.rs [hunk 18]
```diff
@@ -1,3 +1,204 @@
...
+
+// bits for rocket
+#![feature(plugin)]
+#![plugin(rocket_codegen)]
+
+/// #Alchemist
+/// Alchemist_Server is a API server for determing what package should be installed for a distro
+/// based on input. Made for Alchemist, however it should be availibe to use to anyone/script.
+///
+/// The API can be reached at ``https://alchemist.rs/api/v0/`` (you can replace this url if you
```
Comment (SirDoctors) [inline alchemist_server/src/main.rs]: I figured ``/v0/`` was appropriate for a "unstable" API

File: alchemist_server/src/main.rs [hunk 19]
```diff
@@ -1,3 +1,204 @@
...
+/// ## Endpoints
+/// ### Index
+/// ``GET /``
+/// returns the API gretting response.
+/// Example Response:
+///
+/// ```json
+/// {
+/// "msg": "Alchemist public API server",
+/// "version": "0.0.1"
```
Comment (SirDoctors) [inline alchemist_server/src/main.rs]: Sure since we are already following a naming standard for API versions to ensure client/server interopablity between versions.

File: alchemist_server/src/main.rs (hunk 1 above)
Comment (SirDoctors) [inline alchemist_server/src/main.rs]: analytics, sorry. My thought is to build a table in the DB for well known client names to see what is using the API  & version for well.. version checking.

File: alchemist_server/src/main.rs (hunk 2 above)
Comment (SirDoctors) [inline alchemist_server/src/main.rs]: Yep, must have missed that.

File: alchemist_server/src/main.rs (hunk 4 above)
Comment (SirDoctors) [inline alchemist_server/src/main.rs]: Already noticed and fixed locally, will be in the next batch

File: alchemist_server/src/main.rs (hunk 3 above)
Comment (SirDoctors) [inline alchemist_server/src/main.rs]: Nope, will remove.

File: alchemist_server/src/main.rs (hunk 5 above)
Comment (SirDoctors) [inline alchemist_server/src/main.rs]: For empty responses? Like if there is no package to send?

File: alchemist_server/src/main.rs (hunk 6 above)
Comment (SirDoctors) [inline alchemist_server/src/main.rs]: ohh fun, will try it.

File: alchemist_server/src/models.rs (hunk 8 above)
Comment (SirDoctors) [inline alchemist_server/src/models.rs]: That is what I named it in the SQL query, but it sounds good, will change references to that.

File: alchemist_server/src/main.rs (hunk 7 above)
Comment (SirDoctors) [inline alchemist_server/src/main.rs]: Yep, you can mix and match as you please, rust couldn't care as long as it all returns.

File: alchemist_server/src/main.rs [hunk 20]
```diff
@@ -1,3 +1,204 @@
...
+/// }
+///
+/// ### Version
+/// ``GET /version``
+/// Returns Alchemist server & client versions
+/// Example Response:
+/// ```json
+/// {
+/// "client": "0.0.4"
+/// "server": "0.0.1"
```
Comment (SirDoctors) [inline alchemist_server/src/main.rs]: This is actually supposed to be returned from what the github version is at, rather than what the server version is at

File: alchemist_server/.env (hunk 9 above)
Comment (mgattozzi) [inline alchemist_server/.env]: Awesome. A configuration file would be great since it'll help in the future if we want to add options

File: alchemist_server/alchemy/arch.rs (hunk 10 above)
Comment (mgattozzi) [inline alchemist_server/alchemy/arch.rs]: Ah now I remember.

File: alchemist_server/migrations/20170317144416_packages/up.sql (hunk 11 above)
Comment (mgattozzi) [inline alchemist_server/migrations/20170317144416_packages/up.sql]: Awesome. Sounds good to me

File: alchemist_server/migrations/20170317144416_packages/up.sql (hunk 12 above)
Comment (mgattozzi) [inline alchemist_server/migrations/20170317144416_packages/up.sql]: Great!

File: alchemist_server/src/db.rs (hunk 13 above)
Comment (mgattozzi) [inline alchemist_server/src/db.rs]: Yay!

File: alchemist_server/src/db.rs (hunk 14 above)
Comment (mgattozzi) [inline alchemist_server/src/db.rs]: Okay got it.

File: alchemist_server/src/db.rs (hunk 15 above)
Comment (mgattozzi) [inline alchemist_server/src/db.rs]: Okay cool! If you can change it to that it would be great!

File: alchemist_server/src/db.rs (hunk 17 above)
Comment (mgattozzi) [inline alchemist_server/src/db.rs]: I don't think you do. Serde should deserialize an empty vec to [] otherwise it would be [""]

File: alchemist_server/src/main.rs (hunk 18 above)
Comment (mgattozzi) [inline alchemist_server/src/main.rs]: Yeah for now and then when we do a proper release we can bump it up to 1

File: alchemist_server/src/main.rs (hunk 19 above)
Comment (mgattozzi) [inline alchemist_server/src/main.rs]: Sweet thanks!

File: alchemist_server/src/main.rs (hunk 20 above)
Comment (mgattozzi) [inline alchemist_server/src/main.rs]: Meaning?

File: alchemist_server/src/main.rs (hunk 1 above)
Comment (mgattozzi) [inline alchemist_server/src/main.rs]: Oh okay got it

File: alchemist_server/src/main.rs (hunk 4 above)
Comment (mgattozzi) [inline alchemist_server/src/main.rs]: Sweet!

File: alchemist_server/src/main.rs (hunk 3 above)
Comment (mgattozzi) [inline alchemist_server/src/main.rs]: Awesome!

File: alchemist_server/src/main.rs (hunk 5 above)
Comment (mgattozzi) [inline alchemist_server/src/main.rs]: Yeah exactly!

Comment (mgattozzi) [inline alchemist_server/src/main.rs]: Or actually it should just send an empty vec back if there's nothing right?

File: alchemist_server/src/main.rs (hunk 6 above)
Comment (mgattozzi) [inline alchemist_server/src/main.rs]: Did that work out?

File: alchemist_server/src/main.rs (hunk 7 above)
Comment (mgattozzi) [inline alchemist_server/src/main.rs]: Wow I'm really impressed. Normally I'd be doing that logic in the match but now I can do it as part of the match!

File: alchemist_server/src/models.rs (hunk 8 above)
Comment (mgattozzi) [inline alchemist_server/src/models.rs]: Cool thanks!

File: alchemist_server/src/main.rs (hunk 20 above)
Comment (SirDoctors) [inline alchemist_server/src/main.rs]: It pulls the version from github to return, rather than getting it from anywhere else

File: alchemist_server/src/main.rs (hunk 5 above)
Comment (SirDoctors) [inline alchemist_server/src/main.rs]: currently Its setup to send back the string "none" but an empty vec would probably be better.

File: alchemist_server/src/main.rs (hunk 6 above)
Comment (SirDoctors) [inline alchemist_server/src/main.rs]: Nope, so I'm just rewrote everything to take vectors instead of hashmaps (the better way anyhow)

File: alchemist_server/src/main.rs (hunk 20 above)
Comment (mgattozzi) [inline alchemist_server/src/main.rs]: Curious, is their a reason you were thinking the functionality should be included?

File: alchemist_server/src/main.rs (hunk 5 above)
Comment (mgattozzi) [inline alchemist_server/src/main.rs]: I think so as well since it'll be an empty vec on deserialization

File: alchemist_server/src/main.rs (hunk 20 above)
Comment (SirDoctors) [inline alchemist_server/src/main.rs]: At the time the thought was for the convenience of apps using the API, but this has already been removed following the API standardization.

File: alchemist_server/src/main.rs (hunk 6 above)
Comment (mgattozzi) [inline alchemist_server/src/main.rs]: Got it, whenever you finish making the changes push em and I'll take a look again

Comment (SirDoctors) [inline alchemist_server/src/main.rs]: Sounds good!

**Comment by SirDoctors** [comment]
I was just rebasing the branch to be up to date with master, no actual changes

**Comment by SirDoctors** [comment]
I pushed some new changes, I want to get #79 working before merging this into master... however all code not related to that should be good for review.

**Comment by SirDoctors** [comment]
Looks like travis borked its self again, I'll push a commit to fix that when I get the chance. Also going to note that #79 is coming along pretty good, hopefully should be done in a few days.

File: alchemist_server/.env (hunk 9 above)
Comment (mgattozzi) [inline alchemist_server/.env]: File is still here

File: alchemist_server/migrations/20170317144416_packages/up.sql [hunk 21]
```diff
@@ -0,0 +1,14 @@
...
+    arch VARCHAR NOT NULL,
+    ubuntu VARCHAR NOT NULL,
+    mint VARCHAR NOT NULL,
+    debain VARCHAR NOT NULL,
+    gentoo VARCHAR NOT NULL,
+    void VARCHAR NOT NULL,
+    mac VARCHAR NOT NULL,
+    freebsd VARCHAR NOT NULL,
+    netbsd VARCHAR NOT NULL,
+    openbsd VARCHAR NOT NULL
```
Comment (mgattozzi) [inline alchemist_server/migrations/20170317144416_packages/up.sql]: Sweet all lower case now

File: alchemist_server/src/db.rs [hunk 22]
```diff
@@ -0,0 +1,115 @@
...
+
+    let mut output: Vec<Package> = Vec::new();
+
+    for i in input_packages {
+        // While this might look like O(n^2) complexity
+        // it's more closer to O(n) since most querys only
+        // return one or 2 results really. Still this seemed
+        // to be the only way to implement it well
+        let results = packages
+            .filter(arch.eq(&i)
```
Comment (mgattozzi) [inline alchemist_server/src/db.rs]: I still need to ask my DB coworker about this but this is fine for now!

File: alchemist_server/src/db.rs [hunk 23]
```diff
@@ -0,0 +1,115 @@
...
+    let mut pac_converted: Vec<String> = Vec::new();
+
+    // All querys will either be a string or '' in the db allowing us to
+    // use is_empty()
+    // Finds out what distro is used and inserts the proper conversions
+    // into the HashSet to be returned from the function
+    match distro {
+        &Distro::Ubuntu => {
+            for i in results {
+                if !i.ubuntu.is_empty() {
```
Comment (mgattozzi) [inline alchemist_server/src/db.rs]: I feel like we could do a map and filter here, but the logic is sound. I'd prefer that for readability and my own personal tastes because of functional programming.

File: alchemist_server/src/main.rs [hunk 24]
```diff
@@ -1,11 +1,74 @@
...
+#![cfg_attr(feature = "dev", plugin(clippy))]
+#![cfg_attr(feature = "dev", plugin)]
 
+// bits for rocket
+#![feature(plugin, custom_attribute, attr_literals)]
+#![plugin(rocket_codegen)]
+
+//! Alchemist_Server
+//! Alchemist_Server is a API server for determing what package should be installed for a distro
+//! based on input. Made for Alchemist, however it should be availibe to use to anyone/script.
```
Comment (mgattozzi) [inline alchemist_server/src/main.rs]: > Made for Alchemist, however it should be availibe to use to anyone/script.

Could you change this to:

> While it is made for Alchemist, the API is available for use in other programs or scripts that require package conversion.

File: alchemist_server/src/main.rs [hunk 25]
```diff
@@ -1,11 +1,74 @@
...
+//!
+//! The API can be reached at ``https://alchemist.rs/api/v0/`` (you can replace this url if you
```
Comment (mgattozzi) [inline alchemist_server/src/main.rs]: I'll need to register an alchemist.rs domain at some point.

File: alchemist_server/src/main.rs [hunk 26]
```diff
@@ -1,11 +1,74 @@
...
+//! "msg": "Alchemist public API server",
+//! }
+//!
+//! ### Package
+//! ``POST /package``
+//! Takes a JSON POST with four variables ``package``, ``distro``, ``client``, ``client_version``.
+//! ``Package`` should be what the user specified for the package to search for.
+//! ``distro`` is the distro that the user is currently using.
+//! ``client``[ ``name`` & ``version`` ] are for current client accessing the API, used for analytics & to
+//! check if alchemist is up to date, they are needed, however you can put whatever you want in
```
Comment (mgattozzi) [inline alchemist_server/src/main.rs]: > they are needed, however you can put whatever you want in there (doesn't need to be capital) however we'd prefer if you used real client names based on your application.

Can we change this to:

> This field is also required. While you can use whatever you want for the name and version we'd prefer using something that lets us know what tools are accessing the API.

File: alchemist_server/src/main.rs [hunk 27]
```diff
@@ -1,11 +1,74 @@
...
+//! there (doesn't need to be capital) however we'd prefer if you used real client names based on
+//! your application.
+//! Example POST to ``/package``:
+//! ```json
+//! {
+//! "package": ["sudo"],
+//! "distro": "Arch",
+//! "client": {
+//!     "name": "Alchemist",
```
Comment (mgattozzi) [inline alchemist_server/src/main.rs]: Is there stuff checking for this to act differently if it's Alchemist because of different version conflicts?

File: alchemist_server/src/main.rs [hunk 28]
```diff
@@ -1,11 +1,74 @@
...
+//! ``distro`` is the distro that the user is currently using.
+//! ``client``[ ``name`` & ``version`` ] are for current client accessing the API, used for analytics & to
+//! check if alchemist is up to date, they are needed, however you can put whatever you want in
+//! there (doesn't need to be capital) however we'd prefer if you used real client names based on
+//! your application.
+//! Example POST to ``/package``:
+//! ```json
+//! {
+//! "package": ["sudo"],
+//! "distro": "Arch",
```
Comment (mgattozzi) [inline alchemist_server/src/main.rs]: Should this be arch now?

File: alchemist_server/src/main.rs [hunk 29]
```diff
@@ -1,11 +1,74 @@
...
+//! returns the API gretting response.
+//! Example Response:
+//!
+//! ```json
+//! {
+//! "msg": "Alchemist public API server",
+//! }
+//!
+//! ### Package
+//! ``POST /package``
```
Comment (mgattozzi) [inline alchemist_server/src/main.rs]: Since we can put one or more let's make it packages

Comment (mgattozzi) [inline alchemist_server/src/main.rs]: And update the other docs and stuff accordingly

File: alchemist_server/src/models.rs [hunk 30]
```diff
@@ -1,24 +1,18 @@
 /// Struct for a row in the Alchemist package table
-#[derive(PartialEq,PartialOrd,Eq,Ord,Hash,Debug, Queryable)]
+#[derive(PartialEq,PartialOrd,Eq,Ord,Hash,Debug,Queryable)]
```
Comment (mgattozzi) [inline alchemist_server/src/models.rs]: Style wise mind putting a space after each `,`?

File: alchemist_server/src/server_routes.rs [hunk 31]
```diff
@@ -0,0 +1,89 @@
...
+#[post("/package", format = "application/json", data = "<pkgincoming>")]
+pub fn package(pkgincoming: Result<JSON<PkgIncoming>, SerdeError>)
+               -> Result<JSON<Package>, String> {
+    match pkgincoming {
+        Err(why) => {
+            Err(json!({
+                          "error": format!("There was an error parsing the JSON: {:?}", why),
+                          "note": "You probably messed up the syntax, check the docs @ https://alchemist.rs to ensure your sending a valid JSON request."
+                      })
+                        .to_string())
```
Comment (mgattozzi) [inline alchemist_server/src/server_routes.rs]: Could you just put this with the previous line?

File: alchemist_server/src/server_routes.rs [hunk 32]
```diff
@@ -0,0 +1,89 @@
...
+
+// The package post, allows you to POST and get the correct package for your picture
+#[post("/package", format = "application/json", data = "<pkgincoming>")]
+pub fn package(pkgincoming: Result<JSON<PkgIncoming>, SerdeError>)
+               -> Result<JSON<Package>, String> {
+    match pkgincoming {
+        Err(why) => {
+            Err(json!({
+                          "error": format!("There was an error parsing the JSON: {:?}", why),
+                          "note": "You probably messed up the syntax, check the docs @ https://alchemist.rs to ensure your sending a valid JSON request."
```
Comment (mgattozzi) [inline alchemist_server/src/server_routes.rs]: Could you make this multiline so that it stays within 80 chars for the line? Adding a `\` at the end will help break it up and escape the newline char.

File: alchemist_server/src/server_routes.rs [hunk 33]
```diff
@@ -0,0 +1,89 @@
...
+        Some(ref content) if !content.is_json() => {
+            format!("<p>This server only supports JSON requests, not '{}'.</p>",
+                    content)
+        }
+        _ => {
+            json!({
+                      "error": format!("Sorry, '{}' is an invalid path!", request.uri()),
+                      "note": "Try the docs @ https://alchemist.rs for valid paths."
+                  })
+                    .to_string()
```
Comment (mgattozzi) [inline alchemist_server/src/server_routes.rs]: Same as before just put it on the above line

File: alchemist_server/tests/web_server_request.rs [hunk 34]
```diff
@@ -0,0 +1,68 @@
...
+    }
+
+fn rockets() -> rocket::Rocket {
+    rocket::ignite()
+        .mount("/api/v0/", routes![index, package])
+        .catch(errors![not_found])
+}
+
+#[test]
+fn bad_get() {
```
Comment (mgattozzi) [inline alchemist_server/tests/web_server_request.rs]: Awesome you set up a test suite!

File: alchemist_server/tests/web_server_request.rs [hunk 35]
```diff
@@ -0,0 +1,68 @@
...
+    });
+}
+
+#[test]
+fn good_post_on_package() {
+    let rockets = rockets();
+    // Checks "/package" exists and returns the correct value
+    let req = MockRequest::new(Post, "/api/v0/package")
+            .header(ContentType::JSON)
+            .body(r#"{ "package": [ "sudo" ], "distro": "Arch", "client": { "name": "Alchemist", "version": "0.0.4" }}"#);
```
Comment (mgattozzi) [inline alchemist_server/tests/web_server_request.rs]: Same as earlier can you make this multiline?

File: alchemist_server/src/main.rs (hunk 28 above)
Comment (SirDoctors) [inline alchemist_server/src/main.rs]: No as the enum that its check by uses Capital distro names.

File: alchemist_server/src/main.rs (hunk 27 above)
Comment (SirDoctors) [inline alchemist_server/src/main.rs]: No, it would work better as a separate request (perhaps "/alchemist"?) for that functionality, that way even if you are using an older version the request still gets you the packages you need.

Something we could do though is add "Alchemist" to the list of packages to install if a new version exists in the users repository's.

File: alchemist_server/src/models.rs (hunk 30 above)
Comment (SirDoctors) [inline alchemist_server/src/models.rs]: I can try, I believe (though I could be wrong) that rustfmt formally removed all the spaces, will give it a shot)

File: alchemist_server/src/server_routes.rs (hunk 32 above)
Comment (SirDoctors) [inline alchemist_server/src/server_routes.rs]: yeah I'm working on some rustfmt config changes that limit the string length to 80 chars. after/when I do that I'll do that.

File: alchemist_server/tests/web_server_request.rs (hunk 35 above)
Comment (SirDoctors) [inline alchemist_server/tests/web_server_request.rs]: I will give it a shot, however I don't feel that this needs it, as it's supposed to just be a json string to send to the server when its being tested.

**Comment by SirDoctors** [comment]
I don't want to merge this in until I have #79 done as part of this merge, I feel like that is a important part of the server's core functionality.

File: alchemist_server/src/db.rs (hunk 23 above)
Comment (SirDoctors) [inline alchemist_server/src/db.rs]: Alright I'll look into that change, that function is mostly copy paste from the old client code.

File: alchemist_server/src/main.rs (hunk 24 above)
Comment (SirDoctors) [inline alchemist_server/src/main.rs]: done.

File: alchemist_server/src/server_routes.rs (hunk 31 above)
Comment (SirDoctors) [inline alchemist_server/src/server_routes.rs]: that's a rustfmt change, however I agree it should be on the previous line in this case. I'll look into having rustfmt ignore that line change.

Comment (SirDoctors) [inline alchemist_server/src/server_routes.rs]: #[rustfmt_skip] Seems to be the solution to that problem.

**Comment by mgattozzi** [comment]
Ahh I see. Normally if someone puts in a PR and notes why it shouldn't be merged yet then nothing will happen until it's done. That's why I had removed it but good to know you're waiting till you close #79 with it!

File: alchemist_server/src/db.rs (hunk 23 above)
Comment (mgattozzi) [inline alchemist_server/src/db.rs]: Yeah that's from how I would do things a long time ago. You could try:

```rust
results.into_iter()
       .filter(|package| {
           !package.ubuntu.is_empty()
       })
```

File: alchemist_server/src/main.rs (hunk 24 above)
Comment (mgattozzi) [inline alchemist_server/src/main.rs]: Awesome thanks!

File: alchemist_server/src/main.rs (hunk 28 above)
Comment (mgattozzi) [inline alchemist_server/src/main.rs]: Perfect thanks for clearing that up!

File: alchemist_server/src/main.rs (hunk 27 above)
Comment (mgattozzi) [inline alchemist_server/src/main.rs]: Hmm I'd rather it just send back something saying hey update your stuff rather than implicitly updating it without them expecting it too. This is fine then as it.

File: alchemist_server/src/models.rs (hunk 30 above)
Comment (mgattozzi) [inline alchemist_server/src/models.rs]: Huh but it didn't do it in another place. I think you can configure rustfmt to skip different types of things with a config file in the top level.

File: alchemist_server/tests/web_server_request.rs (hunk 35 above)
Comment (mgattozzi) [inline alchemist_server/tests/web_server_request.rs]: It's more a style thing. Over 80 chars tends to wrap the code online and you have to use slider to see the whole thing rather than just seeing it all at once.

File: alchemist_server/src/models.rs (hunk 30 above)
Comment (SirDoctors) [inline alchemist_server/src/models.rs]: You can, I have a commit I'm sitting on to send in with a change or to for rustfmt.

File: alchemist_server/src/server_routes.rs (hunk 31 above)
Comment (SirDoctors) [inline alchemist_server/src/server_routes.rs]: It moves all . operators like .to_string() to a separate line.

File: alchemist_server/tests/web_server_request.rs (hunk 35 above)
Comment (SirDoctors) [inline alchemist_server/tests/web_server_request.rs]: Ahh I see what you mean, alright will move that to a new line then.

**Comment by SirDoctors** [comment]
Yep, it's being worked on, but if it takes too long and everything else is fixed feel free to merge it. It shouldn't take that long though.

**Comment by mgattozzi** [comment]
There's no rush your stuff is independent from the other code so this is fine!

File: alchemist_server/src/server_routes.rs (hunk 31 above)
Comment (mgattozzi) [inline alchemist_server/src/server_routes.rs]: Yeah I like that but some edge cases end up looking ugly. Ah well.

File: alchemist_server/tests/web_server_request.rs (hunk 35 above)
Comment (mgattozzi) [inline alchemist_server/tests/web_server_request.rs]: I really appreciate that thanks! :D

File: alchemist_server/src/db.rs (hunk 23 above)
Comment (SirDoctors) [inline alchemist_server/src/db.rs]: Was hoping I'd be able to figure it out, but I'm totally lost on how to fix that.
```rust
error[E0308]: mismatched types
  --> src/db.rs:77:13
   |
77 |               results
   |  _____________^ starting here...
78 | |                 .into_iter()
79 | |                 .filter(|package| !package.ubuntu.is_empty())
   | |_____________________________________________________________^ ...ending here: expected (), found struct `std::iter::Filter`
   |
   = note: expected type `()`
              found type `std::iter::Filter<std::vec::IntoIter<models::Package>, [closure@src/db.rs:79:25: 79:61]>`
```

File: alchemist_server/src/main.rs (hunk 29 above)
Comment (SirDoctors) [inline alchemist_server/src/main.rs]: done

File: alchemist_server/src/main.rs (hunk 26 above)
Comment (SirDoctors) [inline alchemist_server/src/main.rs]: done

**Comment by mgattozzi** [comment]
@SirDoctors I was wrong with the code I gave you for the filter. You needed to do a filter/map/collect and I forgot that was what was needed. I've updated the code above for you.

**Comment by mgattozzi** [comment]
@SirDoctors just letting you know I saw this and will get to reviewing when I can!

File: alchemist/src/main.rs [hunk 36]
```diff
@@ -120,7 +121,7 @@ fn run() -> Result<()> {
         }
 
         if let Some(p) = matches.values_of("install") {
-            let package_inputs: HashSet<&str> = p.collect();
+            let package_inputs : HashSet<&str> = p.collect();
```
Comment (mgattozzi) [inline alchemist/src/main.rs]: I'm really not a fan of all the ` :` in this pr because of rustfmt. I think it's harder to read.

File: rustfmt.toml [hunk 37]
```diff
@@ -1,2 +1,10 @@
-format_strings = false
+format_strings = true
 reorder_imports = true
+max_width = 150
```
Comment (mgattozzi) [inline rustfmt.toml]: Could you make this 80 please?

File: rustfmt.toml [hunk 38]
```diff
@@ -1,2 +1,10 @@
...
+report_todo = "Always"
+report_fixme = "Always"
+comment_width = 100
```
Comment (mgattozzi) [inline rustfmt.toml]: Could this also be 80 please?

File: alchemist/src/main.rs (hunk 36 above)
Comment (SirDoctors) [inline alchemist/src/main.rs]: I'm not either, going to remove that and amend the commit.

File: rustfmt.toml (hunk 37 above)
Comment (SirDoctors) [inline rustfmt.toml]: That's a rustfmt setting that controls the width of the line it checks, it will just fail to work if set to 80 for any lines even a char over 80. (if set) Which is why it is set high, to give rustfmt a chance to make it shorter.

File: rustfmt.toml (hunk 38 above)
Comment (SirDoctors) [inline rustfmt.toml]: Yeah, pretty sure I left it at that while I was testing the configuration, I'll remove that at the same time I fix the `` : `` in functions and what not.

Comment (mgattozzi) [inline rustfmt.toml]: Okay cool thanks! Sorry been a bit busy and I'm at RustFest now.

File: rustfmt.toml (hunk 37 above)
Comment (mgattozzi) [inline rustfmt.toml]: Got it that's fine

**Comment by SirDoctors** [comment]
Going to update this pr for serde 1.0

**Comment by mgattozzi** [comment]
@SirDoctors sounds good to me!
